### PR TITLE
Add a detailed tutorial for creation of the Widget and The Renderer 

### PR DESCRIPTION
**Pull request by stuartdd2**
As discussed. I have added a step by step guide to the creation of a simple widget and it's renderer

### CODE REVIEW

File: tutorial/write-custom-widget.md [hunk 1]
```diff
@@ -69,9 +69,9 @@ type WidgetRenderer interface {
 As you can see the `Layout(Size)` and `MinSize()` functions are similar to the
 `fyne.Layout` interface, without the `[]fyne.CanvasObject` parameter - this is because a widget does need to be laid out but it controls which objects will be included.
 
-The `Refresh()` method is triggered when the widget this renderer draws has changed or if the theme is altered.
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:72]: Why the change from method to function? These refer to calls on an object instance which is typically referred to as method to differentiate from the functions that are not attached to a type.

File: tutorial/write-custom-widget.md [hunk 2]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
 are well suited to that location but user state would not be. A widget that is
 hidden may have it's renderer destroyed and if it is shown again the new renderer
 must be able to reflect the same widget state.
+
+### Example Widget
+
+The following covers in detail the creation of a simple Widget and it's Renderer.
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:113]: I don't think the apostrophe belongs there?

File: tutorial/write-custom-widget.md [hunk 3]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+
+#### Create a Widget
+
+So the widget requires the`widget.BaseWidget`. For example:
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:117]: Could it be better introduced? Dropping right in with "So" seems unclear.
Also there is a space missing

File: tutorial/write-custom-widget.md [hunk 4]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+
+```go
+type MyWidget struct {
+	widget.BaseWidget           // Inherit from BaseWidget
+	minSize           fyne.Size // The minimum size of the widget
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:122]: Storing this internally could be a distraction from the example - it is not a normal thing to do

File: tutorial/write-custom-widget.md [hunk 5]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+	text              string	// The text to display in the widget
+}
+```
+
+To construct the widget create a New<widgetname> function. The example we are using is called MyWidget so by convention the following constructor function name would be used:
+
+```go
+func NewMyWidget(W, H float32, text string) *MyWidget {
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:130]: Width and Height are controlled by the parent container and should not be passed in to a constructor.
The text parameter is enough for this example anyhow.

File: tutorial/write-custom-widget.md [hunk 6]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+	w := &MyWidget{ 							 // Initialise the widget storage
+		text:    text,                           // Save the text
+		minSize: fyne.Size{Width: W, Height: H}, // Set the size
+	}
+	w.ExtendBaseWidget(w) // Required to initialise the base widget
+	return w
+}
+```
+
+We need to override the MinSize() function so we can return the minimum size of our widget.
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:140]: This is not really true - the default `MinSize` method will look up the renderer information, which is completed later in the example.

File: tutorial/write-custom-widget.md [hunk 7]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+}
+```
+
+Note : It is not necessary to override the Resize function (shown above) but if you need to change the Resize behaviour of your widget and you add a ```Resize(s fyne.Size)``` function then you **must** include the ```w.BaseWidget.Resize(s)```. in the code.
+
+The final function to add is the ```CreateRenderer() fyne.WidgetRenderer``` function:
+
+```go
+func (w *MyWidget) CreateRenderer() fyne.WidgetRenderer {
+	return newMyWidgetRenderer(w.minSize, w.text)
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:157]: This is not best form - you are capturing the state of the widget when creating the renderer - this means that changes in state will never be reflected in the UI.
Instead just pass the widget in, so the renderer can access whichever state is needed.

File: tutorial/write-custom-widget.md [hunk 8]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+}
+```
+
+**Warning**: Do not keep a reference to the widget renderer in the widget (or any other part of your application). The fyne application caches the renderer and may destroy it at any time.
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:161]: These 3 paragraphs could probably all be in one - take up less space.

File: tutorial/write-custom-widget.md [hunk 9]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+
+It is ok for the widget renderer to refer to the widget but not the other way round.
+
+The widget is now almost complete except for the ```newMyWidgetRenderer()``` function which creates your new widget renderer. 
+
+#### Create a Widget Renderer
+
+The widget renderer is how we tell the fyne application what to draw and how to lay it out.
+
+The widget renderer has it's own state and functions (see ```fyne.WidgetRenderer``` interface) but this should only be concerned with drawing (rendering) the widget.
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:173]: I don't really understand this sentence, can you explain what it is trying to outline?

File: tutorial/write-custom-widget.md [hunk 10]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+
+The renderer is responsible for:
+
+1) The objects that are to be drawn. These are returned by the ```Objects() []fyne.CanvasObject``` function. Notice that you can return any object as long as it implements the ```fyne.CanvasObject``` interface.
+2) Each object returned by the Objects() function will need to be created by the widget renderer and initialised before returning.
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:178]: This is not really another point in the "responsible for" list - more like a note against point 1.

File: tutorial/write-custom-widget.md [hunk 11]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+3) The fyne application will then call the renderers Layout() function. This is where the size and position of each object is detirmined.
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:179]: As you are talking about resposibilities this sentence should probably be the other way around - the WidgetRenderer will position and size all items within the space passed in to the Layout method.

File: tutorial/write-custom-widget.md [hunk 12]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+
+Our example renderer could be stored in the following struct:
+
+```
+type myWidgetRenderer struct {
+	minSize fyne.Size // The size of the widget
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:185]: This is not size, it is minimum - as noted below storing it is probably unwise

File: tutorial/write-custom-widget.md [hunk 13]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+	rect    *canvas.Rectangle
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:186]: Why not use a semantic name like "background" or "bg" so it is clearer what these variables are.
Repetition is rather anti-go.

File: tutorial/write-custom-widget.md [hunk 14]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+	text    *canvas.Text
+}
+```
+
+As you can see it is about canvas objects that the fyne application will render. A lower case name is used as the renderer code is private and only created by the Widget. 
+
+The constructor is the place to pass in properties like minimum size, text values and other data.
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:193]: Again, it should not pass in graphical state like minimum size, this is defined by the renderer

File: tutorial/write-custom-widget.md [hunk 15]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+func newMyWidgetRenderer(size fyne.Size, text string) *myWidgetRenderer {
+	return &myWidgetRenderer{
+		minSize: size,  // For the MinSize() function
+		rect:    canvas.NewRectangle(theme.BackgroundColor()),
+		text:    canvas.NewText(text, theme.ForegroundColor()),
+	}
+}
+```
+
+Notice that the size and position of the canvas objects are not defined here, only colours, text and other properties. It might be appropriate to pass in a reference to the widget instead of lots of different parameters. This is left to your design requirements.
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:205]: I think we should recommend passing the object parameter - as mentioned above the renderer will show stale data otherwise.

File: tutorial/write-custom-widget.md [hunk 16]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+	// Resize the background
+	r.rect.Resize(s)
+}
+```
+
+We finish with the remaining functions of the ```fyne.WidgetRenderer``` interface
+
+```go
+func (r *myWidgetRenderer) Destroy() {} // Called when the renderer is destroyed
+func (r *myWidgetRenderer) Refresh() {} // This does not seem to be used!
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:232]: Any time your object state is altered refresh will be called.
In this example it would be important to set the text in the `canvas.Text` to match the data inside the widget.

File: tutorial/write-custom-widget.md [hunk 17]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+func (r *myWidgetRenderer) MinSize() fyne.Size {
+	return r.minSize // Return the minimum size of the widget
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:234]: You should use measureText, as above. We must always return the true minimum size of this widget so that the UI can adapt appropriately to all screens.

File: tutorial/write-custom-widget.md [hunk 18]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+	w := app.NewWindow("My Widget")
+	mw := NewMyWidget(100, 100, "Widget")
+	w.SetContent(mw)
+	w.ShowAndRun()
+}
+```
+
+If you stretch the widget the text will remain centered.
+
+#### Next Steps
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:262]: I think this section of code belongs in a different tutorial doesn't it?

File: tutorial/write-custom-widget.md [hunk 19]
```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
...
+
+See also:
+
+```
+SecondaryTappable
+DoubleTappable
+Disableable
+```
+
+Note: Implementing the DoubleTappable interface introduces a slight delay of about 0.5 seconds before the Tapped() function is called. This is to allow for the next tap to be recognised.
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:284]: The value changes based on driver, probably best not to mention a specific

**Review by andydotxyz — CHANGES_REQUESTED** [review]
Thanks for this. A few things inline which are not quite accurate that could be tidied up.
Also check the indenting of comments - a few are off, plus we use "method" to refer to a function attached to a type - please match that paradigm.

File: tutorial/write-custom-widget.md (hunk 1 above)
Comment (stuartdd2) [reply tutorial/write-custom-widget.md:72]: I will revert to method. This is a throwback to java 'methods', in go the keyword is 'func'. 
Is it correct to say that static functions (e.g. constructors) are 'functions' and that functions bound to a struct are 'methods'?

Just curious!

Comment (Jacalz) [reply tutorial/write-custom-widget.md:72]: Yes that is correct. Anything that is a value or pointer receiver (basically anything that you call with your created type.something()) are methods. Everything else is functions.

Comment (stuartdd) [reply tutorial/write-custom-widget.md:72]: I have not abandoned this. I am on holiday. I will return to this as soon as I am able.

Comment (stuartdd2) [reply tutorial/write-custom-widget.md:72]: I have made changes but I am not sure what happens next. Do I resolve the conversations as I  or the person who started the conversation. 

I have also been pushing multiple commits. I do this because I use two different laptops to make the changes and I need to sync between them. One in my home office and one I take with me. I am not sure if this is the correct approach.

The changes are considerable. I assume another review will be done and a new set of conversations will be generated. How is this process managed. The corporate environment was somewhat different.

Regards
Stuart

Comment (andydotxyz) [reply tutorial/write-custom-widget.md:72]: We prefer small commits where each has separate area/intent.
And the person raising the question usually resolves it so they can confirm they are happy with the resolution.
You will find comments appear here if anything more needs changed - GitHub notifies collaborators when new commits are pushed.

File: tutorial/write-custom-widget.md [hunk 20]
```diff
@@ -107,3 +107,184 @@ Take care not to keep any important state in your renderer - animation tickers
...
+Go provides an simple mechanism for inheriting behaviour from other types. Adding an Anonymous type as a field in the widget structure means that the widget inherits the behaviour of that Anonymous type.
+
+This also means that your widget now implements all of the public methods and properties define by the  `widget.BaseWidget` type.
+
+We add an Anonymous `widget.BaseWidget` as follows:
+
+```go
+type MyWidget struct {
+	widget.BaseWidget           // Inherit from BaseWidget
+	text              string	// The text to display in the widget
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:128]: The alignment here is untidy - I think it is because there is a tab character after `string` so it is not lining up - it should be just spaces after content (tabs indent at the beginning only).

File: tutorial/write-custom-widget.md (hunk 7 above)
Comment (andydotxyz) [reply tutorial/write-custom-widget.md:157]: You have fixed the `newMyWidgetRenderer` code below, but this function call is still passing the old parameters

File: tutorial/write-custom-widget.md [hunk 21]
```diff
@@ -107,3 +107,184 @@ Take care not to keep any important state in your renderer - animation tickers
...
+}
+```
+
+Notice that the size and position of the canvas objects are not defined here, only initial colours and text properties.
+
+The Objects() method should return the canvas objects in a list as follows:
+
+```go
+func (r *myWidgetRenderer) Objects() []fyne.CanvasObject {
+    // The order is critical, rect is drawn first then text
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:211]: Here you have used spaces instead of a tab so the alignment is not correct

File: tutorial/write-custom-widget.md [hunk 22]
```diff
@@ -107,3 +107,184 @@ Take care not to keep any important state in your renderer - animation tickers
...
+Each object returned will need to be sized and positioned in the Layout() method as follows:
+
+```go
+func (r *myWidgetRenderer) Layout(s fyne.Size) {
+	ts := fyne.MeasureText(r.text.Text, r.text.TextSize, r.text.TextStyle)
+	r.text.Move(fyne.Position{X: (s.Width - ts.Width) / 2, Y: (s.Height - ts.Height) / 2})
+	r.background.Resize(s)
+}
+```
+
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:225]: I wonder if this bulleted list would benefit from a short sentence leading in?
Maybe "As you can see in the code sample, our `Layout` method arranges are objects by doing:".

File: tutorial/write-custom-widget.md [hunk 23]
```diff
@@ -107,3 +107,184 @@ Take care not to keep any important state in your renderer - animation tickers
...
+If ```myWidgetRenderer``` does NOT implement the ```fyne.WidgetRenderer``` interface this line will be in error!
+
+#### Use your Widget
+
+```go
+func main() {
+	app := app.New()
+	w := app.NewWindow("My Widget")
+	mw := NewMyWidget(100, 100, "Widget")
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:284]: This usage is the old method signature I think

**Review by andydotxyz — CHANGES_REQUESTED** [review]
Looking great thanks, just a few things missed - comments inline

**Comment by Jacalz** [comment]
I'll try to have a look at this later today or tomorrow as well :)

**Review by andydotxyz — APPROVED** [review]
Great, thanks

**Comment by andydotxyz** [comment]
@Jacalz did you want to look this over too?

**Comment by Jacalz** [comment]
> @Jacalz did you want to look this over too?

Yeah. I'll have a look later today.

File: tutorial/write-custom-widget.md [hunk 24]
```diff
@@ -107,3 +107,186 @@ Take care not to keep any important state in your renderer - animation tickers
...
+
+### Example Widget
+
+The following covers in detail the creation of a simple Widget and its Renderer.
+
+#### Create a Widget
+
+As described above, we need to inherit from the `widget.BaseWidget` type.
+
+Go provides an simple mechanism for inheriting behaviour from other types. Adding an Anonymous type as a field (without a name) in the widget structure means that the widget inherits the behaviour of that Anonymous type.
```
Comment (Jacalz) [inline tutorial/write-custom-widget.md:119]:
```suggestion
Go provides a simple mechanism for inheriting behaviour from other types. Adding an Anonymous type as a field (without a name) in the widget structure means that the widget inherits the behaviour of that Anonymous type.
```

File: tutorial/write-custom-widget.md [hunk 25]
```diff
@@ -107,3 +107,186 @@ Take care not to keep any important state in your renderer - animation tickers
...
+func NewMyWidget(text string) *MyWidget {
+	w := &MyWidget{
+		text: text,
+	}
+	w.ExtendBaseWidget(w)
+	return w
+}
+```
+
+Note : It is not necessary to override the Resize method (shown below) but if you need to change the Resize behaviour of your widget and you override the ```Resize(s fyne.Size)``` method then you **must** include the ```w.BaseWidget.Resize(s)```. in the code.
```
Comment (Jacalz) [inline tutorial/write-custom-widget.md:144]:
```suggestion
Note: It is not necessary to override the Resize method (shown below) but if you need to change the Resize behaviour of your widget and you override the `Resize(s fyne.Size)` method then you **must** include the `w.BaseWidget.Resize(s)`. in the code.
```
This is a bit of a nitpick, but inline code blocks should only use one of those "things". This also applies to other parts in the code.

**Review by Jacalz — CHANGES_REQUESTED** [review]
This is looking very good. Massive thanks :)

I mostly just had some small fixes and suggestions to be a bit more correct in some places. The only thing I would suggest apart from that is to work a bit more with the end. It seems a bit abrupt. It could do with an image of the widget and maybe some text to make it feel a bit more like the tutorial is finished and not just "If you stretch the widget the text will remain centered.".

File: tutorial/write-custom-widget.md (hunk 25 above)
Comment (andydotxyz) [reply tutorial/write-custom-widget.md:144]: That is a good catch, sorry I missed that one

**Comment by iMartyn** [comment]
Hi, as someone who is trying to implement a custom widget for fyne, I really welcome improvements to the documentation, but at least the way I understand things better is to see a full example of working code I can pull apart and change things.  Perhaps there's room here to put a full code example at the bottom?  Something that you can just put in your repo and compiles but does nothing (significant), that people can use as a starting point.

**Review by andydotxyz — APPROVED** [review]

**Review by Jacalz — CHANGES_REQUESTED** [review]
> The only thing I would suggest apart from that is to work a bit more with the end. It seems a bit abrupt. It could do with an image of the widget and maybe some text to make it feel a bit more like the tutorial is finished and not just "If you stretch the widget the text will remain centered.".

Sorry but this part is still missing from my review. I also think the suggestion from @iMartyn is a good one.

**Comment by stuartdd2** [comment]
I agree. I will add an image and the complete widget code.

File: tutorial/write-custom-widget.md [hunk 26]
```diff
@@ -235,7 +235,9 @@ We do this in the renderer because only the renderer knows the size of the canva
 
 ```go
 func (r *myWidgetRenderer) MinSize() fyne.Size {
-	return fyne.MeasureText(r.text.Text, r.text.TextSize, r.text.TextStyle)
+	// Measure the size of the text so we can calculate an indent.
+	ts := fyne.MeasureText(r.text.Text, r.text.TextSize, r.text.TextStyle)
+	return fyne.NewSize(ts.Width+20, ts.Height+20)
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:240]: I'm not sure about this magic 20, what is the reason?
If you want it to be padded from the outside consider using a multiple of `theme.Padding()` like other widgets (see https://developer.fyne.io/architecture/widgets)

File: tutorial/write-custom-widget.md [hunk 27]
```diff
@@ -288,5 +290,8 @@ func main() {
 	w.ShowAndRun()
 }
 ```
+The widget should look like this:
+	
+![widgetDemoScreenSmall](https://user-images.githubusercontent.com/94919638/145826575-53ae12fc-e6d0-4976-9037-dbdaaea9f4de.png)
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:295]: Please don't reference private user images. If you want to add the image, add it in the PR to this repository.

File: tutorial/write-custom-widget.md [hunk 28]
```diff
@@ -288,5 +290,8 @@ func main() {
...
 
-If you stretch the widget the text will remain centered.
+	The widget can be resized and the text will remain centered.
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:297]: Is this indent intentional?

**Review by andydotxyz — CHANGES_REQUESTED** [review]

File: tutorial/write-custom-widget.md [hunk 29]
```diff
@@ -235,9 +233,9 @@ We do this in the renderer because only the renderer knows the size of the canva
 
 ```go
 func (r *myWidgetRenderer) MinSize() fyne.Size {
-	// Measure the size of the text so we can calculate an indent.
-	ts := fyne.MeasureText(r.text.Text, r.text.TextSize, r.text.TextStyle)
-	return fyne.NewSize(ts.Width+20, ts.Height+20)
+    // Measure the size of the text so we can calculate an indent.
+    ts := fyne.MeasureText(r.text.Text, r.text.TextSize, r.text.TextStyle)
+    return fyne.NewSize(ts.Width+theme.Padding()*5, ts.Height+theme.Padding()*5)
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:238]: Try 4, not 5 - that way it would match widget sizes

File: tutorial/write-custom-widget.md [hunk 30]
```diff
@@ -283,15 +281,178 @@ If `myWidgetRenderer` does NOT implement the `fyne.WidgetRenderer` interface thi
...
-![widgetDemoScreenSmall](https://user-images.githubusercontent.com/94919638/145826575-53ae12fc-e6d0-4976-9037-dbdaaea9f4de.png)
 
-	The widget can be resized and the text will remain centered.
+![](../images/widgets/customWidgetSimple.png)
+
+The widget can be resized and the text will remain centered.
+
+### The complete Widget
+
+Note that some minor additions have be made to the code below:
```
Comment (andydotxyz) [inline tutorial/write-custom-widget.md:300]: Why add more code? The complete demo should be as minimal as the code above, otherwise you are adding more items that are not explained.
Additionally this code uses deprecated APIs - we should not encourage developers to override user theme settings.

**Review by andydotxyz — CHANGES_REQUESTED** [review]
This commit seems to have broken all the indentation moving from tab to 4 space.
Also it has introduced lots more code - can we keep it focused on the documented APIs please?

File: tutorial/write-custom-widget.md (hunk 28 above)
Comment (andydotxyz) [reply tutorial/write-custom-widget.md]: > I have included a section of code to demonstrate how the widget works when the theme is changed however the code `theme.DarkTheme()` has been depreciated. I have looked for the equivalent without success. Usually a recommendation is given in the depreciation warning but I could not see it!

There is no recommended replacement because doing this is not recommended.
User settings should be respected and the `SetTheme` API is now focused on apps that provide a custom theme.

**Comment by andydotxyz** [comment]
It's a shame that this was so close but now needs more work.
The additional of lots more code has raised questions and need for more change. Why not just include the code that was already documented as a summary at the bottom? It would be much more logical than adding new functionality as a demo app.

**Comment by Jacalz** [comment]
Sorry, this needs to rebased on master after all of the restructuring that @andydotxyz did lately.

**Comment by andydotxyz** [comment]
Oh yes sorry about that - the new file location is `extend/custom-widget.md`

**Comment by stuartdd2** [comment]
Do I need to create a new pull request and resolve conflicts to align with fyne-io:master.

**Comment by andydotxyz** [comment]
Whichever is easier for you - new PR or merge and resolve.
Sorry for the extra work.

**Comment by stuartdd2** [comment]
I have merged but still need to fix some indents and things that got lost in the process

**Comment by andydotxyz** [comment]
> As a result I see the shorter tabbing in VSCode and xed. The wider tabbing if I copy from any formatted source and paste in to the markdown.
> 
> This means that the tabbing is inconsistent with the original (shorter) tabbing in the original document.

How it looks is not always the way to tell if it is correct. The code should be indented using tabs (Go standard) not spaces. This may be 8 or 4 wide in your editor, it doesn't matter.
GitHub (or other git client) will show you when you've changed tabs to spaces by mistake.

**Comment by stuartdd2** [comment]
Sorry I am very confused. **I have used gofmt to output to the console and to a file.** I assume that is what you meant by 'Go standard'.

When I look at those files in VSCode and xed they are rendered indented 4. I believe VSCode uses 'go fmt' on save.

When copied from VSCode or xed or directly from the **Console**  in to a markdown code block or directly to git markdown code block they render as indented 8. Its driving me nuts!

I will do some more experiments to see what I am doing wrong.

**Comment by stuartdd2** [comment]
After some probing it seems that for github my default indentation was 8. 
I have corrected this in my settings.appearance but it has not fixed the issue for the file.

It I click to edit the file in github I can see the indentation is still 8 even though when I go to settings.appearance to check it says 4.

While editing I can change the indentation to 4 and it all looks good but it does not stick. I have set it to 4, made a small change to the file so I could commit it (github tab is not a property of the file) and when I return to the file my tab is set to 8 again. My settings.appearance is unaffected.

I will continue investigating.

**Comment by andydotxyz** [comment]
I am sure this is a question of spaces vs tabs. In Go we do not use spaces to indent, we use tabs.
In some languages you could choose 2 vs 4 vs 8 spaces, and that is saved into the file - with tabs it is 1 and the user configuration changes how it looks.
When GitHub shows the indentation change it is normally because a tab character has been replaced by (for example 4) spaces.

**Comment by andydotxyz** [comment]
Closing as 2 years stale. We would still love this contribution, feel free to push (once the indent is fixed) to the new docs.fyne.io repo.
